Updates to the Docker requirements for v2

DavidLiedle opened the pull request:
This PR brings numerous discussions into written form, breaking the requirements into 4 distinct topics:

1. Docker Compose
1. Deployment
1. Data Migration
1. CLI Wrapper

DavidLiedle committed:
Updates to the Docker requirements for v2

andrerom commented on DX/Docker/Compose.md:
```diff
@@ -0,0 +1,35 @@
+# v2: Docker Compose Requirements
+
+## Objective
+To provide a default Docker Compose file or files, ready to use out-of-the-box.
```
It's not just docker compose, blue print or whatever we call it also involves docker files (containers), and ways to configure eZ Platform from docker (env variables) which needs changes on eZ Platform itself _(like in 2.0 branch: https://github.com/ezsystems/ezplatform/pull/167/files#diff-bf0e70bcef1a5d5b2f87289220a51108R10)_

DavidLiedle replied:
Updating per feedback, adding mention of Dockerfiles

andrerom commented on DX/Docker/Compose.md:
```diff
@@ -0,0 +1,35 @@
+The feature SHOULD:
+1. Be able to reconfigure eZ Platform via environment variables for all possible
+  server configurations, such as:
+  1. Single/cluster setup
```
rending gets wrong here

DavidLiedle committed:
Fixed bad markdown indentation

2 spaces changed to 4 for intended nesting of sub-items.

rolandbenedetti commented on DX/Docker/Wrapper.md:
```diff
@@ -0,0 +1,50 @@
+
+### Firm
+The solution MUST:
+1. Run _inside_ of a Docker container, to unify environments across all systems,
```
It seems we have here an issue. On one end, we have the opportunity to build a PHP tool that would deliver most of what we try to achieve, but add the requirement of having PHP running on the command line of the developer environment. On the other, we can still look for a full Docker based solution, but it seems we don't really know how to get there, nor what would be the effort or the technical solution. Pragmatically, I'd like to go for the first route, to quickly enable developers to deploy on their machines. I'd like to revise this requirement expect if we can show a fully Docker based solution that is requiring same level of effort/risk. I have looked at other applications available on Docker hub (such as Drupal or Wordpress) but each time, for anything complex, they end up asking the user to manually do Docker Compose, so I'm not sure about the value.

My vote: be pragmatic and go for a 1st step which require Docker and PHP on CLI.
Insights here highly appreciated.

Plopix commented:
Hello here,

I think there is no debate on the PHP client approach for the wrapper. PHP is the eZ language of the eZ Community. It will bring more power than bash especially with the Symfony Console and Process components.(YAML component to manage the docker-compose.yml etc.)

Of course, this PHP client MUST not require non-default php-extension (like intl or others). I would say php5.6.x+.

I have concerned about:
> Run inside of a Docker container, to unify environments across all systems, and thus equally usable on Unix, Linux, Windows, and macOS machines.

Did you check the feasibility of this? How I see the way to do it would be to mount the Docker socket inside a PHP container to bring PHP on a system that does not have it. And then inside this container, we would run the PHP Client that would create the stack of the eZ Platform (running docker-compose finally).
I am not sure that this "solution" is correct. I think it is not supposed to work like that I don't this the value.

Something like that would be required:
`docker run -t --rm -v /var/run/docker.sock:/var/run/docker.sock  -w /code -v $(pwd):/code php:5.6-alpine php SOMETING`

The only reason to add this complexity (or this generic way to run the PHP Client) would be because it is hard to install PHP.
=> Which is not to my opinion.

Then comes the Windows case.
- First Docker Toolbox (that installs Virtualbox etc.) is required before a specific build of Windows 10.
- There is not TTY on Windows 

Then the Docker in Docker could be the only solution (unless we find a way to manage interactivity in PHP). For command in the docker stack that will require input.

Maybe the Windows use case must be addressed differently?

xrow commented on DX/Docker/Compose.md:
```diff
@@ -0,0 +1,35 @@
+    1. Search engine using db/solr/es
+    1. Http cache using Symfony or Varnish
+    1. Session stored on disk or in shared backend like memcached/redis/db
+1. Be able to initiate with a database and binary dump
```
https://image.slidesharecdn.com/bjorndiedingrunningezplatformonkubernetes-161011145757/95/running-ez-platform-on-kubernetes-presented-by-bjrn-dieding-at-ez-conference-2016-19-638.jpg?cb=1476197957

This is how we do it in Kubernetes. As the next step we created a Autodiscover Bundle.

xrow commented on DX/Docker/README.md:
```diff
@@ -1,91 +1,88 @@
+cluster and scale, there are numerous entry points into implementation of these
+options. For eZ Platform, we will begin with a focus on the following:
+
+1. Development / Staging / Production containers, composed by eZ and editable by
```
I would say containers as of today are useless for development. If one would show me how I would compare towards vagrant.

xrow commented on DX/Docker/Migration.md:
```diff
@@ -0,0 +1,29 @@
+
+### Preferred
+The feature SHOULD:
+1. Be able to do this operation from dev to prod, and from prod back to dev
```
Highly desired a matter of the webserver and or init system being able to forward env to the app.

DavidLiedle replied:
Adding `Be able to forward ENV variables to the desired app`

xrow commented on DX/Docker/Migration.md:
```diff
@@ -0,0 +1,29 @@
+1. ...to share database schema changes with my team
+1. ...to migrate production from the old schema to the new one
+1. ...to enjoy identical environments between developers, staging, and production
+
```
I want to shrink my 50 GB DB

DavidLiedle replied:
Adding `...to optimize the size of my DB`

xrow commented on DX/Docker/Wrapper.md:
```diff
@@ -0,0 +1,50 @@
+  README.md).
+1. Be modular, following the UNIX approach to piping collections of commands, in
+  which each command _does one thing_.
+1. Keep and expose logs to the developer
```
Send the logs to a log service.

rolandbenedetti committed:
Update the Wrapper requirements (#27)

This is an update with Roland's understanding of where we are (adding Björn's suggestion in the meantime).
- Feasibility of a fully Docker based solution is not sure.
- PHP requirement seems very reasonable to most of us for this first step.
Thanks for you review, we'd like to move forward quickly :-)

DavidLiedle commented:
New PR from @rolandbenedetti has been merged in, revising the Wrapper requirements. I will make additional updates from the feedback requests from @andrerom and @xrow

DavidLiedle committed:
Updated text per feedback in discussions

rolandbenedetti commented:
This looks good to me.
+1
